### ISSUE TITLE
[v0.3.1] gRPC calls stop being proxied / hang (under investigation)

### ISSUE THREAD

**Issue by bourquep** [runconduit/conduit#584]
_This is a new bug with 0.3.1, as the 2 services involved in the following description had the conduit 0.3.0 injected until today and this scenario has been working fine since 0.3.0._

I am still trying to figure out what is going on, so I will update this issue as I go.

## Scenario

I have a `reports` service which monitors a RabbitMQ queue for tasks. These tasks represent a report that must be generated and emailed to a user. 

Every friday at 8pm eastern, a few thousand tasks are added to this rabbit queue, and the number of pods in the `reports` service is bumped to 24 to process these weekly email reports.

When processing a task, the `report` service invokes the `generator` service via a single gRPC method call. I have between 6 and 12 pods running for the `generator` service.

It looks like this:

`RabbitMQ <-amqp- reports -gRPC-> generator`

Until today, Conduit 0.3.0 proxy was injected in both `reports` and `generator` services, and everything was dandy. 

Tonight, as I type this, my `reports` pods keep getting stuck and I am not sure why.

I have scaled my `reports` deployment to a single pod to help me diagnose, and am skipping outbound port 5672 (RabbitMQ) from the `reports` service to rule out any potential issues with Conduit vs amqp.

## Observations

Here are my observations so far. For clarity, I will use `client` to designate the `reports` service, and `server` to designate the `generator` service:

* With 1 client pod and 20 server pods, the single client pod hangs after processing 22-24 tasks.
* With 1 client pod and 10 server pods, the single client pod hangs after processing ~12 tasks.
* With 1 client pod and 6 server pods, the single client pod hangs after processing 7-10 tasks.
  * By "hang", I mean it gets a task from RabbitMQ but the gRPC call is issued but never delivered to the `generator` deployment. I have logging in the client just before issuing the gRPC call, and logging in the server when receiving a request, and both of those logs become quiet when it "hangs".
  * this behavior is systematic: I kill the client pod, a new one is spun up by k8s, and the new pod's logs exhibit the same behavior: hang after sending 7-10 emails. I have repeated this 10 times with always the same result.
  * the `conduit-proxy` container logs from within that single client pod don't show any error:

```
INFO conduit_proxy using controller at HostAndPort { host: DnsName("proxy-api.conduit.svc.cluster.local"), port: 8086 }
INFO conduit_proxy routing on V4(127.0.0.1:4140)
INFO conduit_proxy proxying on V4(0.0.0.0:4143) to None
INFO conduit_proxy::transport::connect "controller-client", DNS resolved DnsName("proxy-api.conduit.svc.cluster.local") to 10.35.241.37
```

* Then I scaled to 0 client pods and 2 server pods, then scaled back to 1 client pod.
  * Everything was running fine
* Then I scaled to 6 server pods, everything still fine (so I was having the same 1 client/6 servers configuration as all my failed attempts described above, but no hang at all now).
* So I scaled to 12 client pods, everything fine
* I ended up with 24 client pods and 12 server pods (the usual setup for a friday evening email blast) and no more problems, I got through the end of the job queue.

## TL;DR

* Let `client_pods = 1`
* Let `server_pods = 6`
* The Nth gRPC call hangs/is not delivered
  * where `N` is just a bit greater than `server_pods`
* This issue was always happening after restarting the client pod
* When I scaled client deployment to 0, effectively removing all pods, and scaled it back to 1, I wasn't able to reproduce the issue anymore, regardless of the number of client and server pods.

## Post-scriptum

I realize that there is not much hard data to analyse here, and that the chances of reproing that bug are very slim. It might not even a problem with Conduit, although everything was ok until today.

I will monitor next friday's email blast to see if it happens again. Is there anything I should do before it starts in order to have better visibility into the problem, if it is to manifest itself again? Such as enabling verbose logging in Conduit (proxy and/or control plane)?

**Comment by hawkw** [comment]
@bourquep Before next Friday, I would re-inject the proxy with `--proxy-log-level=conduit_proxy=trace,hyper=trace,h2=trace,tower_balance=trace`. This will make the proxy's logs quite verbose, but it should hopefully capture any information that could be useful for debugging this issue.

**Comment by hawkw** [comment]
Hi @bourquep, did you observe this behaviour again during the generation of yesterday's email blast?

**Comment by bourquep** [comment]
Yes, I started with 1 client pod consuming the Rabbit queue, and it hung after processing a few items. I scaled the deployment to 0, then scaled back to 1, everything was ok. Scaled it up to 24 (the usual number) and every email was sent successfully.

Unfortunately, I forgot to make the logs more verbose...

**Comment by hawkw** [comment]
Okay, thanks for letting me know. I'm sorry to hear that the issue occurred again. Did you see any out of the ordinary messages in the logs this time? I'm afraid that without more information from the logs, it's pretty difficult to diagnose what's going on here.

**Comment by bourquep** [comment]
Nope, nothing useful in the logs, but I had not enabled verbose logging so that was to be expected.

I will make sure to enable this next friday.

**Comment by olix0r** [comment]
Conduit v0.4.0 will include fixes for some situations where H2 streams are reset. We should continue to debug this, though.

**Comment by bourquep** [comment]
For the record, last friday's email blast went without a hitch with Conduit v0.4.0.

**Comment by hawkw** [comment]
@bourquep that's good to hear! 🙌

**Comment by wmorgan** [comment]
Any objections to closing this issue?
